core/commands/store/upload/ds/datastore.go

```diff
@@ -36,9 +43,30 @@ func Get(d ds.Datastore, key string, m proto.Message) error {
 	return err
 }
 
-func List(d ds.Datastore, prefix string) {
-	d.Query(query.Query{
+func List(d ds.Datastore, prefix string, limit int) ([]proto.Message, error) {
+	ms := make([]proto.Message, 0)
+	if limit <= 0 || limit > limitList {
+		limit = limitList
+	}
+	results, err := d.Query(query.Query{
 		Prefix:  prefix,
 		Filters: []query.Filter{},
+		Limit:   limit,
 	})
+	if err != nil {
+		return nil, err
+	}
+	for nxt, ok := results.NextSync(); ok; nxt, ok = results.NextSync() {
```

Comment:
why not just standard doc suggests:

`for entry := range result.Next() { ... }`